src/gui/sdlmain.cpp

```diff
@@ -773,17 +786,106 @@ static bool LoadGLShaders(const char *src, GLuint *vertex, GLuint *fragment) {
 }
 #endif
 
-Bitu GFX_SetSize(Bitu width,Bitu height,Bitu flags,double scalex,double scaley,GFX_CallBack_t callback) {
+/* Get the maximum area available for drawing: */
+/* TODO: The current implementation will not let the pixel-perfect mode to  */
+/*       to apply even the simplest 1:2 and 2:1 corrections at the original */
+/*       window resolution, therefore: consider a special case for PP mode. */
+static void GetAvailableArea(int *width, int *height)
+{
+	bool   fixed;
+	double par;
+
+	fixed = false;
+	if (sdl.desktop.fullscreen) {
+		if (sdl.desktop.full.fixed) {
+			*width  = sdl.desktop.full.width;
+			*height = sdl.desktop.full.height;
+			fixed   = true;
+		}
+	}
+	else {
+		if (sdl.desktop.window.width > 0) {
+			*width  = sdl.desktop.window.width;
+			*height = sdl.desktop.window.height;
+			fixed   = true;
+		}
+	}
+	if (!fixed) {
+		par = sdl.draw.par;
+		if (par > 1.0)
+			*height = round( *height * par );
+		if (par < 1.0)
+			*width  = round( *width  / par ); 
+	}
+}
+
+// ATT: aspect is the final aspect ratio of the image including its pixel dimensions and PAR
+/*static void GetActualArea( Bit16u av_w, Bit16u av_h, Bit16u *w, Bit16u *h, double aspect )
+{	double as_x, as_y;
+
+	if( aspect > 1.0 )
+	{	as_y =     aspect; as_x = 1.0;  } else
+	{	as_x = 1.0/aspect; as_y = 1.0;  }
+
+	if( av_h / as_y < av_w / as_x )
+	{	*h = av_h; *w = round( (double)av_h / aspect );  } else
+	{	*w = av_w; *h = round( (double)av_w * aspect );  }
+}*/
+
+/* Initialise pixel-perfect mode: */
+static bool InitPp(Bit16u avw, Bit16u avh)
+{
+	bool   ok;
+	double newpar;
+	/* TODO: consider reading apsect importance from the .ini-file */
+	ok = pp_getscale(
+		sdl.draw.width, sdl.draw.height,
+		sdl.draw.par,
+		avw, avh,
+		1.14, /* relative importance of aspect ratio */
+		&sdl.ppscale_x,
+		&sdl.ppscale_y
+	) == 0;
+	if (ok) {
+		newpar = ( double )sdl.ppscale_y / sdl.ppscale_x;
+		LOG_MSG( "Pixel-perfect scaling:\n"
+			"%ix%i (%#4.3g) --[%ix%i]--> %ix%i (%#4.3g)",
+			sdl.draw.width, sdl.draw.height, sdl.draw.par,
+			sdl.ppscale_x,  sdl.ppscale_y,
+			sdl.ppscale_x * sdl.draw.width, sdl.ppscale_y * sdl.draw.height,
+			newpar);
+	}
+	return ok;
+}
+
+Bitu GFX_SetSize(Bitu width,Bitu height,Bitu flags,double scalex,double scaley,GFX_CallBack_t callback,double par) {
+	int avw, avh; /* available width and height */
+	Bitu retFlags = 0;
 	if (sdl.updating)
 		GFX_EndUpdate( 0 );
 
+	sdl.draw.par = par;
 	sdl.draw.width=width;
 	sdl.draw.height=height;
 	sdl.draw.callback=callback;
 	sdl.draw.scalex=scalex;
 	sdl.draw.scaley=scaley;
 
-	Bitu retFlags = 0;
+	sdl.dbl_h = (flags & GFX_DBL_H) > 0;
+	sdl.dbl_w = (flags & GFX_DBL_W) > 0;
+
+	avw = width; avh = height;
+	GetAvailableArea(&avw, &avh);	
+	LOG_MSG( "Input image: %ix%i DW: %i DH: %i PAR: %5.3g",
+		(int)width, (int)height, sdl.dbl_h, sdl.dbl_w, par );
+	LOG_MSG( "Available area: %ix%i", avw, avh );
+	if (sdl.kind == OkPerfect) {
+		if (!InitPp( avw, avh ) ) {
```

Comment:
> line 880 has 2 tabs instead of a tab followed by spaces.

That was intended. It is just a line break according to the [kernel style](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#breaking-long-lines-and-strings). Do you mean that all long invocations must be aligned, instead of simply indented? Here is the result:
~~~
	LOG_MSG("Input image: %ix%i DW: %i DH: %i PAR: %5.3g",
	        (int)width, (int)height, sdl.double_h, sdl.double_w, par);
~~~
Is that what you mean? The other remarks fulfilled.